torchmultimodal/modules/layers/position_embedding.py

```diff
@@ -57,76 +57,66 @@ def __init__(
-        return int(torch.prod(torch.tensor(self.shape)).item())
+    def indices(self) -> Tensor:
+        """Returns broadcasted indices of the data
 
-    @property
-    def decode_idxs(self) -> List:
-        """Indices along the dims of data, e.g., ``(time, height, width)``."""
-        return list(itertools.product(*[range(s) for s in self.shape]))
+        For example::
+
+            >>> pos_emb = BroadcastedPositionEmbedding(shape=(2, 3), embedding_dim=6)
+            >>> pos_emb.indices
+            tensor([[0, 0], [0, 1], [0, 2], [1, 0], [1, 1], [1, 2]])
+        """
+        return torch.cartesian_prod(*[torch.arange(s) for s in self.shape])
 
     def _broadcast(self, i: int) -> Tensor:
         """Broadcasts the ``i``-th embedding matrix ``(self.shape[i], self.embedding_dim // n_dim)`` along the other
         dims of ``self.shape``. The embedding dim is not touched.
 
-                For example::
+        For example::
 
             >>> pos_emb = BroadcastedPositionEmbedding(shape=(2, 4), embedding_dim=6)
             >>> print(pos_emb.embedding["d_0"].shape)
             torch.Size([2, 3])
             >>> pos_emb.embedding["d_0"] = nn.Parameter(torch.tensor([[0., 0., 0.], [0., 0., 1.]]))
             >>> out = pos_emb._broadcast(i=0)
             >>> print(out)
-            tensor([[[[0, 0, 0], [0, 0, 0], [0, 0, 0], [0, 0, 0]],
-                    [[0, 0, 1], [0, 0, 1], [0, 0, 1], [0, 0, 1]]]])
+            tensor([[[0, 0, 0], [0, 0, 0], [0, 0, 0], [0, 0, 0]],
+                    [[0, 0, 1], [0, 0, 1], [0, 0, 1], [0, 0, 1]]])
             >>> print(out.shape)
-            (1, 2, 4, 3)
+            (2, 4, 3)
 
         The input is broadcasted along the second dim ``4`` since it's the ``0``-th embedding constructed w.r.t the
         first dim ``2``.
         """
         emb = self.embedding[f"d_{i}"]
-        # (1, 1, ..., 1, self.shape[i], 1, ..., -1)
+        # (1, ..., 1, self.shape[i], 1, ..., embedding_dim)
         emb = emb.view(
-            1,
             *itertools.repeat(1, i),
             self.shape[i],
             *itertools.repeat(1, (self.n_dim - i - 1)),
             -1,
         )
-        # (1, *self.shape, -1)
-        emb = emb.expand(1, *self.shape, -1)
+        # (*self.shape, embedding_dim)
+        emb = emb.expand(*self.shape, -1)
 
         return emb
 
-    def _decode(
-        self, decode_step: int, embeddings: Tensor, x_shape: Tuple[int, ...]
-    ) -> Tensor:
-        """Returns the embedding vector immediately before the decoding location."""
-        decode_idx = self.decode_idxs[decode_step - 1]
-        embeddings = tensor_slice(
-            embeddings,
-            [0, *decode_idx, 0],
-            [x_shape[0], *itertools.repeat(1, self.n_dim), x_shape[-1]],
-        )
-
-        return embeddings
+    def forward(self, position_ids: Tensor) -> Tensor:
+        if torch.max(position_ids) >= len(self.indices) or torch.min(position_ids) < -1:
```

Comment:
what does a position_id of -1 mean?

Reply:
`-1` means the last position in the position encodings. I'm not sure if `-1` would be something that people use realistically but anything below -1 would definitely break the code (and hence the check).